test/lib/array.jl

```diff
@@ -18,3 +18,46 @@ test_rrule(ZygoteRuleConfig(), x->sum(sin, Diagonal(x)), rand(3); rrule_f=rrule_
     @test gradient(x -> sum([y[2] * y[3] for y in Iterators.product(x, x, x, x)]), [1,2,3,4])[1] ≈ [320, 320, 320, 320]
     @test gradient(x -> sum(y[2] * y[3] for y in Iterators.product(x, x, x, x)), [1,2,3,4])[1] ≈ [320, 320, 320, 320]
 end
+
+@testset "collect" begin
+    @testset "Dict" begin
+        d = Dict(1 => 5, 2 => 6)
+        k = 2
+        i = findfirst(p -> p[1] == k, collect(d))
+        g = gradient(d -> collect(d)[i][2], d)[1]
+        @test g isa Dict{Int64, <:Union{Nothing, Int64}}
+        @test g[k] == 1
+
+        g = gradient(d -> sum(v^2 for (_,v) in collect(d)), d)[1]
+        @test g isa Dict{Int,Int}
+        @test g == Dict(1 => 10, 2 => 12)
+    end
+
+    @testset "NamedTuple" begin
+        t = (a=1, b=2)
+        g = gradient(d -> sum(x^2 for x in collect(d)), t)[1]
+        @test g isa NamedTuple{(:a, :b), Tuple{Float64, Float64}}
+        @test  g == (a = 2, b = 4)
```

Comment:
```suggestion
        @test g === (a = 2.0, b = 4.0)
```
I believe this is sufficient to capture both tests